Add async enumerables (#24776)
dotnet/docs: docs/csharp/language-reference/keywords/yield.md

The sequence returned from an iterator method can be consumed by using a [foreach](../statements/iteration-statements.md#the-foreach-statement) statement or LINQ query. Each iteration of the `foreach` loop calls the iterator method. When a `yield return` statement is reached in the iterator method, `expression` is returned, and the current location in code is retained. Execution is restarted from that location the next time that the iterator function is called.

> When the iterator returns an <xref:System.Collections.Generic.IAsyncEnumerable%601?displayProperty=nameWithType>, that sequence can be consumed asynchronously using an [await foreach](../statements/iteration-statements.md#await-foreach) statement. The iteration of the loop is analogous to the `foreach` statement. The difference os that each iteration may be suspended for an asynchronous operation before returning the expression for the next element.

Reviewer: ```suggestion
When the iterator returns an <xref:System.Collections.Generic.IAsyncEnumerable%601?displayProperty=nameWithType>, that sequence can be consumed asynchronously using an [await foreach](../statements/iteration-statements.md#await-foreach) statement. The iteration of the loop is analogous to the `foreach` statement. The difference is that each iteration may be suspended for an asynchronous operation before returning the expression for the next element.
```

Minor typo.

The declaration of an iterator must meet the following requirements:

> - The return type must be <xref:System.Collections.Generic.IAsyncEnumerable%601>, <xref:System.Collections.Generic.IEnumerable%601>, <xref:System.Collections.IEnumerable>, <xref:System.Collections.Generic.IEnumerator%601>, or <xref:System.Collections.IEnumerator>.

Reviewer: ```suggestion
- The return type must be one of the following types:
  - <xref:System.Collections.Generic.IAsyncEnumerable%601>
  - <xref:System.Collections.Generic.IEnumerable%601>
  - <xref:System.Collections.IEnumerable>
  - <xref:System.Collections.Generic.IEnumerator%601>
  - <xref:System.Collections.IEnumerator>
```

Format as a nested-list for readability.

On each subsequent iteration of the `foreach` loop, the execution of the iterator body continues from where it left off, again stopping when it reaches a `yield return` statement. The `foreach` loop completes when the end of the iterator method or a `yield break` statement is reached.

The following code returns an `IAsyncEnumerable<string>` from an iterator method and then iterates through its elements.

```csharp
IAsyncEnumerable<string> elements = MyAsyncIteratorMethod();
await foreach (string element in elements)
{

>    ...

Reviewer: ```suggestion
   // ...
```

I strive to ensure that code snippets would compile if copied and pasted, I know that `MyAsyncIteratorMethod` doesn't exist, but `...` is even less accurate. I try to leave these as comments.
